C++: Expose SSA definitions from dataflow

MathiasVP opened the pull request:
This PR makes it possible to use the dataflow SSA library outside the internal dataflow directory.

This has been a request by people for many years, but I've pushed back on it since we didn't feel the API was stable enough to publicly expose it. However, the API basically hasn't changed since we switched to the "new" dataflow library more than two years ago. So I think it's pretty safe to say that it's stable enough at this point.

I don't think we need a DCA run for this since:
- No queries actually use these classes at this point, and
- No functionality is changing. We are just exposing already existing classes which we've been using for years.

cc @bdrodes since you've been quite noisy about wanting this.

I did a small drive-by fix while here in commit 1: Schack left a TODO for us in https://github.com/github/codeql/pull/18942 which I've fixed. This makes the predicate identical to Java and C#.

aschackmull commented:
I don't think any other language exposes SSA in terms of data flow nodes. I can see the motivation, though, it's a much shorter path to impact than, say, getting a good SSA on the AST nodes. But it's important to highlight that this is not really pointing in the direction of language api unification (the glorious future).
I don't like the name `DataFlow::Definition` - I think `Ssa` should be in there somewhere in the name: Either `DataFlow::Ssa::Definition` or `DataFlow::SsaDefinition` (same of course goes for the other class names).

MathiasVP commented:
> I don't think any other language exposes SSA in terms of data flow nodes. I can see the motivation, though, it's a much shorter path to impact than, say, getting a good SSA on the AST nodes. But it's important to highlight that this is not really pointing in the direction of language api unification (the glorious future). I don't like the name `DataFlow::Definition` - I think `Ssa` should be in there somewhere in the name: Either `DataFlow::Ssa::Definition` or `DataFlow::SsaDefinition` (same of course goes for the other class names).

I've renamed the classes in de0bb48ec7fbf2891fa2c23f4cea20f01f0ca50b, and modified the API to not expose dataflow nodes (but rather `Instruction`s and `Operand`s) in 889925b638285b7acb0523baef5db8cd7b57de64.

I do agree that we ultimately would like a new abstraction around `(Expr, int)` pairs, but I'll leave that for the future.

I did a drive-by cleanup in 6dacc64515820d954e5eabed261367760901a048. Nothing exciting is happening there (it's just renaming a few files and changing the name of some private import aliases).

aschackmull commented on cpp/ql/lib/semmle/code/cpp/controlflow/SSA.qll:
````diff
@@ -15,6 +15,13 @@ class StandardSsa extends SsaHelper {
+ * library:
+ * ```
+ * import semmle.code.cpp.dataflow.new.DataFlow
+ * // use `DataFlow::Definition`
````
Needs an update to use the `Ssa::` prefix.

MathiasVP replied:
Doh! Good catch!

MathiasVP replied:
Fixed in 0e9286dd3412f0fed412df7ad8341d40058ac956
